### PR TITLE
feat: (Reorg Support) Perform Reorg internally with blocktx statuses

### PR DESCRIPTION
**Pull request by kuba-4chain**
## Description of Changes

Whenever a block that extends the competing chain arrives, with the greatest chainwork, the internal reorg is performed:
1. A common ancestor is found between these two competing chains.
2. All `STALE` blocks from the common ancestor, up to the tip of competing chain, are marked as `LONGEST`
3. All previously `LONGEST` blocks from the common ancestor, up to the previous longest tip are marked as `STALE`.
4. Everything is being saved in db.

An integration test was updated to ensure a proper flow between blocktx processor logic and updates in blocktx store.

## Linked Issues / Tickets

ARCO-177

## Testing Procedure

Describe the tests you've added or any testing steps you've taken.

- [x] I have added new unit tests
- [x] All tests pass locally
- [x] I have tested manually in my local environment

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have updated `CHANGELOG.md` with my changes

### CODE REVIEW

File: internal/blocktx/store/postgresql/get_stale_chain.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
+package postgresql
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/bitcoin-sv/arc/internal/blocktx/blocktx_api"
+)
+
+func (p *PostgreSQL) GetStaleChainBackFromHash(ctx context.Context, hash []byte) ([]*blocktx_api.Block, error) {
```
Comment (boecklim) [inline internal/blocktx/store/postgresql/get_stale_chain.go:10]: Could you add here a comment which explains what this function does?
It looks not easy to understand.

File: internal/blocktx/store/postgresql/postgres_test.go [hunk 2]
```diff
@@ -218,6 +218,78 @@ func TestPostgresDB(t *testing.T) {
 		require.ElementsMatch(t, expectedBlockGaps, blockGaps)
 	})
 
+	t.Run("get longest chain from height", func(t *testing.T) {
+		prepareDb(t, postgresDB.db, "fixtures/get_longest_chain")
+
+		starting_height := uint64(822014)
+		hash0_longest := testutils.RevChainhash(t, "0000000000000000025855b62f4c2e3732dad363a6f2ead94e4657ef96877067")
+		hash1_longest := testutils.RevChainhash(t, "000000000000000003b15d668b54c4b91ae81a86298ee209d9f39fd7a769bcde")
```
Comment (boecklim) [inline internal/blocktx/store/postgresql/postgres_test.go:226]: Variables should be in camelCase

File: internal/blocktx/store/postgresql/postgres_test.go [hunk 3]
```diff
@@ -218,6 +218,78 @@ func TestPostgresDB(t *testing.T) {
...
+			require.Equal(t, expectedStaleHashes[i][:], b.Hash)
+		}
+	})
+
+	t.Run("get stale chain back from hash", func(t *testing.T) {
+		prepareDb(t, postgresDB.db, "fixtures/get_stale_chain")
+
+		hash2_stale := testutils.RevChainhash(t, "00000000000000000659df0d3cf98ebe46931b67117502168418f9dce4e1b4c9")
+		hash3_stale := testutils.RevChainhash(t, "0000000000000000082ec88d757ddaeb0aa87a5d5408b5960f27e7e67312dfe1")
+		hash4_stale := testutils.RevChainhash(t, "000000000000000004bf3e68405b31650559ff28d38a42b5e4f1440a865611ca")
```
Comment (boecklim) [inline internal/blocktx/store/postgresql/postgres_test.go:247]: Variables should be in camelCase

File: internal/blocktx/store/postgresql/postgres_test.go [hunk 4]
```diff
@@ -218,6 +218,78 @@ func TestPostgresDB(t *testing.T) {
...
+			require.Equal(t, expectedStaleHashes[i], b.Hash)
+		}
+	})
+
+	t.Run("update blocks statuses", func(t *testing.T) {
+		prepareDb(t, postgresDB.db, "fixtures/update_blocks_statuses")
+
+		hash2_stale := testutils.RevChainhash(t, "00000000000000000659df0d3cf98ebe46931b67117502168418f9dce4e1b4c9")
+		hash3_stale := testutils.RevChainhash(t, "0000000000000000082ec88d757ddaeb0aa87a5d5408b5960f27e7e67312dfe1")
+		hash4_stale := testutils.RevChainhash(t, "000000000000000004bf3e68405b31650559ff28d38a42b5e4f1440a865611ca")
```
Comment (boecklim) [inline internal/blocktx/store/postgresql/postgres_test.go:269]: Variables should be in camelCase

**Review by boecklim — APPROVED** [review]

File: internal/blocktx/store/postgresql/postgres_test.go (hunk 2 above)
Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/postgres_test.go]: Right, done

File: internal/blocktx/store/postgresql/postgres_test.go (hunk 3 above)
Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/postgres_test.go]: Done

File: internal/blocktx/store/postgresql/postgres_test.go (hunk 4 above)
Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/postgres_test.go]: Done

File: internal/blocktx/store/postgresql/get_stale_chain.go (hunk 1 above)
Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/get_stale_chain.go:22]: I tried to explain it a bit in the comment above the function. Let me know if that makes sense 🙏

Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/get_stale_chain.go:22]: I added one more comment explaining the query a bit more.

But a better explanation of recursive queries can be found here: https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-recursive-query/

File: internal/blocktx/processor.go [hunk 5]
```diff
@@ -447,8 +447,15 @@ func (p *Processor) processBlock(msg *p2p.BlockMessage) error {
 		incomingBlock.Status = blocktx_api.Status_STALE
 
 		if hasGreatestChainwork {
-			// TODO: perform reorg - next ticket
+			p.logger.Info("reorg detected - updating blocks", slog.String("incoming block hash", blockHash.String()), slog.Uint64("height", incomingBlock.Height))
```
Comment (arkadiuszos4chain) [inline internal/blocktx/processor.go:450]:
```suggestion
			p.logger.Info("reorg detected - updating blocks", slog.String("hash", blockHash.String()), slog.Uint64("height", incomingBlock.Height))
```

File: internal/blocktx/processor.go [hunk 6]
```diff
@@ -447,8 +447,15 @@ func (p *Processor) processBlock(msg *p2p.BlockMessage) error {
...
+
 			incomingBlock.Status = blocktx_api.Status_LONGEST
+
+			err := p.performReorg(ctx, incomingBlock)
+			if err != nil {
+				p.logger.Error("unable to perform reorg", slog.String("incoming block hash", blockHash.String()), slog.Uint64("height", incomingBlock.Height), slog.String("err", err.Error()))
```
Comment (arkadiuszos4chain) [inline internal/blocktx/processor.go:456]:
```suggestion
				p.logger.Error("unable to perform reorg", slog.String("hash", blockHash.String()), slog.Uint64("height", incomingBlock.Height), slog.String("err", err.Error()))
```

**Review by arkadiuszos4chain — CHANGES_REQUESTED** [review]

File: internal/blocktx/processor.go [hunk 7]
```diff
@@ -556,6 +563,33 @@ func (p *Processor) hasGreatestChainwork(ctx context.Context, incomingBlock *blo
...
+		lowestHeight = getLowestHeight(staleBlocks)
+	}
+
+	longestBlocks, err := p.store.GetLongestChainFromHeight(ctx, lowestHeight)
+	if err != nil {
+		return err
+	}
+
+	longestBlocksHashes := prepareHashes(longestBlocks)
+	err = p.store.UpdateBlocksStatuses(ctx, longestBlocksHashes, blocktx_api.Status_STALE)
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/blocktx/processor.go]: Good point. I did even better - in one query 🏅

File: internal/blocktx/store/postgresql/get_longest_chain.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+			hash
+		 ,prevhash
+		 ,merkleroot
+		 ,height
+		 ,processed_at
+		 ,orphanedyn
+		 ,status
+		 ,chainwork
+		FROM blocktx.blocks
+		WHERE height >= $1 AND status = $2
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/get_longest_chain.go:22]: Done, migration created

File: internal/blocktx/store/postgresql/get_longest_chain.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+	rows, err := p.db.QueryContext(ctx, q, height, blocktx_api.Status_LONGEST)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var block blocktx_api.Block
+		var processed_at sql.NullString
```
[the remark this answers is not in the document]
Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/get_longest_chain.go]: Done, changed in all occurences

File: internal/blocktx/processor.go (hunk 5 above)
Comment (kuba-4chain) [reply internal/blocktx/processor.go]: Removed

File: internal/blocktx/processor.go (hunk 6 above)
Comment (kuba-4chain) [reply internal/blocktx/processor.go]: Removed

File: internal/blocktx/store/postgresql/migrations/000015_height_index.up.sql [hunk 10]
```diff
@@ -0,0 +1 @@
+CREATE INDEX idx_block_height ON blocktx.blocks(height);
```
Comment (arkadiuszos4chain) [inline internal/blocktx/store/postgresql/migrations/000015_height_index.up.sql:1]:
```suggestion
CREATE INDEX ix_block_height ON blocktx.blocks(height);
```

**Review by arkadiuszos4chain — CHANGES_REQUESTED** [review]

File: internal/blocktx/store/postgresql/migrations/000015_height_index.up.sql (hunk 10 above)
Comment (kuba-4chain) [reply internal/blocktx/store/postgresql/migrations/000015_height_index.up.sql]: Done

**Review by boecklim — APPROVED** [review]

**Pull request merged**